lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -0,0 +1,308 @@
+        }
+
+        /// <summary>
+        /// Specfies the Map or Update task to run on the active context
+        /// </summary>
+        /// <param name="activeContext"></param>
+        public void OnNext(IActiveContext activeContext)
+        {
+            Logger.Log(Level.Info, string.Format("Received Active Context {0}", activeContext.Id));
+
+            if (_groupCommDriver.IsMasterTaskContext(activeContext))
+            {
+                var partialTaskConf =
+                    TangFactory.GetTang()
+                        .NewConfigurationBuilder(new[]
+                        {
+                            TaskConfiguration.ConfigurationModule
+                                .Set(TaskConfiguration.Identifier,
+                                    IMRUConstants.UpdateTaskName)
+                                .Set(TaskConfiguration.Task,
+                                    GenericType<UpdateTaskHost<TMapInput, TMapOutput, TResult>>.Class)
+                                .Build(),
+                            _configurationManager.UpdateFunctionConfiguration
+                        })
+                        .Build();
+
+                _commGroup.AddTask(IMRUConstants.UpdateTaskName);
+                _groupCommTaskStarter.QueueTask(partialTaskConf, activeContext);
+            }
+            else
+            {
+                string taskId;
+
+                if (!_taskIdStack.TryPop(out taskId))
+                {
+                    Logger.Log(Level.Warning, "No task Ids exist for the active context {0}. Disposing the context.",
+                        activeContext.Id);
+                    activeContext.Dispose();
+                    return;
+                }
+
+                var partialTaskConf =
+                    TangFactory.GetTang()
+                        .NewConfigurationBuilder(new[]
+                        {
+                            TaskConfiguration.ConfigurationModule
+                                .Set(TaskConfiguration.Identifier, taskId)
+                                .Set(TaskConfiguration.Task, GenericType<MapTaskHost<TMapInput, TMapOutput>>.Class)
+                                .Build(),
+                            _configurationManager.MapFunctionConfiguration
+                        })
+                        .Build();
+
+                _commGroup.AddTask(taskId);
+                _groupCommTaskStarter.QueueTask(partialTaskConf, activeContext);
+            }
+        }
+
+        /// <summary>
+        /// Specfies what to do when the task is completed
+        /// In this case just disposes off the task
+        /// </summary>
+        /// <param name="completedTask">The link to the completed task</param>
+        public void OnNext(ICompletedTask completedTask)
+        {
+            _completedTasks.Add(completedTask);
+
+            if (_completedTasks.Count != _dataSet.Count + 1) return;
+            
+            foreach (var task in _completedTasks)
```

Comment:
Why do you wait for this till *all* tasks are done? Is this needed for group communications shutdown?